src/expr/src/expr/wrapper/non_strict.rs

```diff
@@ -0,0 +1,130 @@
+
+use crate::error::Result;
+use crate::expr::{Expression, ValueImpl};
+use crate::ExprError;
+
+/// Report an error during evaluation.
+#[auto_impl(Arc)]
+pub trait EvalErrorReport: Clone + Send + Sync {
+    /// Perform the error reporting.
+    ///
+    /// Called when an error occurs during row-level evaluation of a non-strict expression,
+    /// that is, wrapped by [`NonStrict`].
+    fn report(&self, error: ExprError);
+}
+
+/// A dummy implementation that panics when called.
+impl EvalErrorReport for ! {
+    fn report(&self, _error: ExprError) {
+        unreachable!()
+    }
+}
+
+/// A wrapper of [`Expression`] that evaluates in a non-strict way. Basically...
+/// - When an error occurs during chunk-level evaluation, recompute in row-based execution and pad
+///   with NULL for each failed row.
+/// - Report all error occurred during row-level evaluation to the [`EvalErrorReport`].
+pub struct NonStrict<E, R> {
+    inner: E,
+    report: R,
+}
+
+impl<E, R> std::fmt::Debug for NonStrict<E, R>
+where
+    E: std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NonStrict")
+            .field("inner", &self.inner)
+            .field("report", &std::any::type_name::<R>())
+            .finish()
+    }
+}
+
+impl<E, R> NonStrict<E, R>
+where
+    E: Expression,
+    R: EvalErrorReport,
+{
+    pub fn new(inner: E, report: R) -> Self {
+        Self { inner, report }
+    }
+
+    /// Evaluate expression in row-based execution with `eval_row_infallible`.
+    async fn eval_chunk_infallible_by_row(&self, input: &DataChunk) -> ArrayRef {
+        let mut array_builder = self.return_type().create_array_builder(input.capacity());
+        for row in input.rows_with_holes() {
+            if let Some(row) = row {
+                let datum = self.eval_row_infallible(&row.into_owned_row()).await; // TODO: use `Row` trait
+                array_builder.append(&datum);
+            } else {
+                array_builder.append_null();
+            }
+        }
+        array_builder.finish().into()
+    }
+
+    /// Evaluate expression on a single row, report error and return NULL if failed.
+    async fn eval_row_infallible(&self, input: &OwnedRow) -> Datum {
+        match self.inner.eval_row(input).await {
+            Ok(datum) => datum,
+            Err(error) => {
+                self.report.report(error);
+                None // NULL
+            }
+        }
+    }
+}
+
+// TODO: avoid the overhead of extra boxing.
+#[async_trait]
+impl<E, R> Expression for NonStrict<E, R>
+where
+    E: Expression,
+    R: EvalErrorReport,
+{
+    fn return_type(&self) -> DataType {
+        self.inner.return_type()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        Ok(match self.inner.eval(input).await {
+            Ok(array) => array,
+            Err(_e) => self.eval_chunk_infallible_by_row(input).await,
+        })
+    }
```

Comment:
So this function tries `inner.eval` chunk, if failed, eval it row by row again? Is there a better option?

Reply:
Yes.

> Is there a better option?

Maybe. Assuming the evaluation error is rare, the current implementation works fine, so I didn't make any changes in this PR.